qiskit_ibm_runtime/session.py

```diff
@@ -308,6 +312,7 @@ def __exit__(
         exc_tb: Optional[TracebackType],
     ) -> None:
         set_cm_session(None)
+        self.close()
```

Comment:
Can there be a way to control this behavior in case I want to continue accepting jobs on the session even after I've exited the context manager?

My use case is I have some test code using pytest. I define a fixture which creates a session:

```py
@pytest.fixture
def session(service):
    """Creates and yields a Session on the configured backend"""
    session = Session(service=service, backend=os.getenv('IBM_CLOUD_BACKEND'))
    yield session
    # Make sure that we always close the session so there isn't idle time on the backend.
    LOG.info(f'closing session {session.session_id}')
    session.close()
```

I have a utility function to submit jobs, optionally against a session (from the fixture):

```py
def run_job(service, session=None, wait=True):
    """Runs a Sampler job

    :param service: The QiskitRuntimeService instance to use to run the job
    :param session: Optional session to use for the job. If provided the
    session is left open.
    :param wait: If true, waits for the job to complete.
    :returns: RuntimeJob
    """
    options = Options(
        resilience_level=0,
        execution=dict(shots=1)
    )
    if session:
        with session:
            sampler = Sampler(options=options)
            job = sampler.run(ReferenceCircuits.bell())
            LOG.info(f'created job: {job.job_id()}; session: {job.session_id}; backend: {get_job_backend(job)}')
    else:
        # Only get the backend if one is specified otherwise Qiskit returns
        # the first available which is not what we want if we're trying to
        # let the scheduler pick the backend from the default resource set.
        if 'IBM_CLOUD_BACKEND' in os.environ:
            backend = service.backend(os.getenv('IBM_CLOUD_BACKEND'))
        else:
            backend = None
        sampler = Sampler(backend=backend, options=options)
        job = sampler.run(ReferenceCircuits.bell())
        LOG.info(f'created job: {job.job_id()}; backend: {get_job_backend(job)}')

    if wait:
        wait_for_final_state(job)
    return job
```

And then I have a test which submits jobs against the session before explicitly closing it:

```py
def test_close_session_with_queued_jobs(service, session):
    with session:
        jobs = []
        # concurrent_jobs=3 (for AoT) so create 4 jobs so we're sure to have one queued
        for _ in range(4):
            job = run_job(service, session=session, wait=False)
            jobs.append(job)

        # Immediately explicitly close the session before at least the last job can run.
        LOG.info(f'closing session {session.session_id}')
        session.close()
```

With the change proposed here, when the `with session:` exits in `run_job`, trying to submit another job against the same session will fail.

It'd be nice if there were a way to control that behavior to opt out of it if you know what you're doing, so make setting `accepting_jobs=False` on exit the default, but be able to override that if you have a case for it.

Reply:
@jyu00 thoughts on this? We could maybe have another method like `session.open` that sets `accepting_jobs` to true.

Reply:
That feels a bit clunky because I'd have to call it every time I submit a job to the session. I was thinking more like a flag you could set on the session to control the behavior when the context manager exits.

Reply:
@mriedem can you help me better understand the reason for the change you're asking for? Does the use case also exist outside of the tests you're describing?

Reply:
> in case I want to continue accepting jobs on the session even after I've exited the context manager?

The point of a context manager is to clean up when the context ends. In your case, you should not use a context manager and just pass `session` directly to the primitive, e.g. `Sampler(session=session, options=options)`.

Reply:
> In your case, you should not use a context manager and just pass `session` directly to the primitive, e.g. `Sampler(session=session, options=options)`.

This is what I was wondering, if I should be using `session.run` or something else - sounds like I should be using `Sampler`. I figured there were multiple options. So I'll change to do that.